// @ldapjs/messages/index.js
'use strict'

module.exports = {
  // Base objects.
  LdapMessage: require('./lib/ldap-message'),
  LdapResult: require('./lib/ldap-result'),

  // Request objects.
  AbandonRequest: require('./lib/messages/abandon-request'),
  AddRequest: require('./lib/messages/add-request'),
  BindRequest: require('./lib/messages/bind-request'),
  CompareRequest: require('./lib/messages/compare-request'),
  DeleteRequest: require('./lib/messages/delete-request'),
  ExtensionRequest: require('./lib/messages/extension-request'),
  ModifyRequest: require('./lib/messages/modify-request'),
  ModifyDnRequest: require('./lib/messages/modifydn-request'),
  SearchRequest: require('./lib/messages/search-request'),
  UnbindRequest: require('./lib/messages/unbind-request'),

  // Response objects.
  AbandonResponse: require('./lib/messages/abandon-response'),
  AddResponse: require('./lib/messages/add-response'),
  BindResponse: require('./lib/messages/bind-response'),
  CompareResponse: require('./lib/messages/compare-response'),
  DeleteResponse: require('./lib/messages/delete-response'),
  ExtensionResponse: require('./lib/messages/extension-response'),
  ModifyResponse: require('./lib/messages/modify-response'),
  ModifyDnResponse: require('./lib/messages/modifydn-response'),

  // Search request messages.
  SearchResultEntry: require('./lib/messages/search-result-entry'),
  SearchResultReference: require('./lib/messages/search-result-reference'),
  SearchResultDone: require('./lib/messages/search-result-done'),

  // Specific extension response implementations.
  PasswordModifyResponse: require('./lib/messages/extension-responses/password-modify'),
  WhoAmIResponse: require('./lib/messages/extension-responses/who-am-i'),

  // Miscellaneous objects.
  IntermediateResponse: require('./lib/messages/intermediate-response')
}
